seed/seed20.go

```diff
@@ -96,32 +94,74 @@ type seed20 struct {
 	essentialSnapsNum int
 }
 
-// Copy implement Copier interface.
-func (s *seed20) Copy(seedDir string, label string, tm timings.Measurer) (err error) {
+func shouldCopySnap(target *Snap, model *asserts.Model, oc *OptionalContainers) bool {
+	if oc == nil {
+		return true
+	}
+
+	if target.Essential {
+		return true
+	}
+
+	for _, sn := range model.AllSnaps() {
+		if sn.Name != target.SnapName() {
+			continue
+		}
+
+		if sn.Presence == "required" {
+			return true
+		}
+	}
+
+	return strutil.ListContains(oc.Snaps, target.SnapName())
+}
+
+func shouldCopyComponent(target Component, snapName string, model *asserts.Model, oc *OptionalContainers) bool {
+	if oc == nil {
+		return true
+	}
+
+	for _, sn := range model.AllSnaps() {
```

Comment:
would it be worth to make a name -> ModelSnap map before end?

Reply:
we are going over AllSnaps in these helpers that are called multiple times, I'm wondering whether taking AllSnaps and turning into map[string]*ModelSnap map and passing that to the helper would be helpful?